### DISCUSSION TITLE
Assert. FATAL 16897 handled. Core dump. Aborted.

### DISCUSSION THREAD

**Discussion by kitty99099** [nasa/fprime#2898] [Q&A] [answered]
Hello!
Based on this project https://github.com/nasa/fprime/discussions/2889. I am going to use runhandler with rategroup 1Hz to send all of my telemetry (more than 50) so that all the data could be shown on the GDS. Although it can be really shown some data on the GDS for a while, it turned out to be core dump like below.

`Assert: "/MyProject/build-fprime-automatic-aarch64-linux/Components/MyComponent/MyComponentComponentAc.c`
`FATAL 16897 handled.`
`Exiting with abort signal and core dump file.`
`Aborted`

However, when I changed my number of telemetry sent a time to much less, the problem didn't happened or happened until a much longer time. I'm curious if this problem related to the buffer? I'm not sure how to find out the fundamental issue from these informations above. Thank you for your reading. I look forward to any useful response.

**Comment by thomas-bc** [comment]
https://github.com/nasa/fprime/blob/dfc655f7892802e5ecfcd9e9a9c133c76ef2aa59/config/FpConfig.h#L232

Making this config value larger in your config will allow the assert string to be longer, and we'll be able to see where the assertion is happening in the file, and therefore what is causing it.

**Reply by kitty99099** [reply]
Hey @thomas-bc ,
Thank you for your fast respond again. After I make that larger, the assertion can be seen fully as  below.
`Assert: "/home/tt/Fprime_Tensor_ADCS/build-fprime-automatic-aarch64-linux/Components/MyComponent/MyComponentComponentAc.cpp:1105" 8`

1. Then I checked this line. The result is as below.
![image](https://github.com/user-attachments/assets/a58e936a-d040-47d1-bd23-7c8a013b68ab)

2. It seems like `qstatus != QUEUE_OK`, so I look for the `send()` function in `line 1103` above. The result is as below.
![image](https://github.com/user-attachments/assets/4747235c-fbaa-44c5-9702-fa6c9620f9d6)

3. Then I look for the `send` funciton in the `return` above. The important parts of Queue.hpp and Queue.cpp are respectively as below.
![image](https://github.com/user-attachments/assets/b936b90c-503b-4d29-b315-ab1bb2dd0ed7)
![image](https://github.com/user-attachments/assets/b24892c4-ee7a-48e7-9ebe-dc2dbd6ea93a)

4. I added a line to print the what  was returned by the `send()` function in 2. . I ran the file, and got a number `8`, which means the `QueueStatus` is `QUEUE_FULL` based on the Queue.hpp in 3. .
![image](https://github.com/user-attachments/assets/41db3308-36a4-4f1f-a96a-fb8c81925474)

Now I will keep trying how to solve the `QUEUE_FULL`  problem. If I have any progress, I will post that to let you know. On the other hand, if you have any insight, I will appreciate your further respond!

**Reply by tbkevin** [reply]
I have more than 300 channels. I can't recall if that exactly error appeared to me, but for my case I had to increase "FW_COM_BUFFER_MAX_SIZE"  found in "FpConfig.h", since this value will be check in the next assert, on an autogenerated cpp file:
"static_assert((700 <= (FW_COM_BUFFER_MAX_SIZE - Fw::Time::SERIALIZED_SIZE - sizeof(FwTlmPacketizeIdType) - sizeof(FwPacketDescriptorType))), "PacketsTooBig");"
That '700' value comes from your packet definition in your deployment, the xml file, 'NameOfYourDeploymentPackets.xml'. In my case I had:
![image](https://github.com/user-attachments/assets/7b39cd11-0503-4861-8d92-83f1c7679c72)
Then i had to modify, in "TlmChanImplCfg.hpp", the next parameter:
![image](https://github.com/user-attachments/assets/359192db-7aea-4267-9062-764720b4c82c)

Hope it helps you!

**Reply by kitty99099** [reply]
Hi @tbkevin , thank you for sharing your case. 
After I figured out that I got the `QUEUE_FULL`problem, I tried to survey and then found there is already a QA sililar to mine: #2825 .

In that QA @LeStarch said: 
> Increasing the active logger component's queue size or reducing the message/event rate is the only way to fix it.

In mine case, I can't reduce my message rate because of the design requirement. Therefore, I tried increasing my queue size. It seemed a bit helpful but not fully useful. When the queue was 100, it would trigger the assertion quickly. When the queue was 1000, it would trigger the assertion not until a longer time. When the queue was 10000, it would trigger the assertion not until a much longer time. All of them finally came out to be `QUEUE_FULL` .

I am curious that: when the queue get longer, shouldn't it be cleared automatically for a certain time just like buffer? I thought it could be fixed if I increase the queue size larger enough so that the queue can be cleared automatically. However, based on my result, it seems like I was wrong. Does anyone have insights?

***I was wrong! After more tests, the assertion time seems doesn't change even though I changed the queue size.
I changed the queue size in instance.fpp:
```
  instance bufferManager: Svc.BufferManager base id 0x4400 \
    queue size 10000 \
    stack size Default.STACK_SIZE \
    priority 100
```
So...I'm more curious now...

**Reply by LeStarch** [reply]
Queue overflows happen when a queue is not serviced fast enough to prevent overflow.  This can be caused by:

1. Threaded component is receiving to many messages
2. Threaded component is processing too slowly
3. Queued component is not calling `doDispatch()` frequently enough

This assert is happening in `MyComponent` so you should be looking at the queue size of that component (no buffer manager as I see above).

**Reply by kitty99099** [reply, marked as answer]
Thank you! @LeStarch , I could'n find where to increase the queue size in the past few days. After your explaination, I found it in the `instance.fpp` as below.
```
  instance MyDeployment: Components.MyComponent base id 0x0F00 \
    queue size Default.QUEUE_SIZE \
    stack size Default.STACK_SIZE \
    priority 95
```
Then I tried to find what `QUEUE_SIZE` is. It is defined on the top of `instance.fpp` also.
```
  module Default {
    constant QUEUE_SIZE = 10
    constant STACK_SIZE = 64 * 1024
  }
```
I changed the QUEUE_SIZE from 10 to 100, and then it is better now.
Thank you for your guidence.
